File: src/main/java/script2/Scope.java
package script2;

import java.util.LinkedList;
import java.util.List;

import org.antlr.v4.runtime.ParserRuleContext;

public class Scope {
    //上一级的Scope
    protected Scope enclosingScope; 

    //下一级的Scope 
    protected List<Scope> children = new LinkedList<Scope>();

    //该Scope中的成员，包括变量、方法、类等。
    protected List<Symbol> symbols = new LinkedList<Symbol>();

    //产生Scope那个AST节点
    protected ParserRuleContext ctx = null;

    public Scope() {
    }

    public Scope(Scope enclosingScope) {
        this.enclosingScope = enclosingScope;
        if (enclosingScope != null) {
            enclosingScope.children.add(this);
        }
    }

    public boolean isRoot() {
        return (enclosingScope == null);
    }


    /**
     * 在下级节点中根据context查找Scope。用于...
     * 
     * @param context
     * @return
     */
    // public Scope findDescendantByContext(ParserRuleContext context){
    // Scope rtn = null;
    // for(Scope scope : children){
    // if (scope.context == context){
    // rtn = scope;
    // break;
    // }
    // }

    // if (rtn == null){
    // for(Scope scope : children){
    // rtn = scope.findDescendantByContext(context);
    // if (rtn != null){
    // break;
    // }
    // }
    // }

    // return rtn;
    // }
}

File: lab/scope/Scope.js
/**
 * Scope.js
 * 测试JavaScript的作用域
 */

var a = 5;
var b = 5;
console.log("1: a=%d b=%d", a, b);

if (a > 0) {
    a = 4;
    console.log("2: a=%d b=%d", a, b);
    var b = 3; //看似声明了一个新变量，其实还是引用的外部变量
    console.log("3: a=%d b=%d", a, b);
}
else {
    var b = 4;
    console.log("4: a=%d b=%d", a, b);
}

console.log("5: a=%d b=%d", a, b);

for (var b = 0; b< 2; b++){
    console.log("6-%d: a=%d b=%d",b, a, b);
}

console.log("7: a=%d b=%d", a, b);